### PR TITLE
Pull chaosDNA to main

### PR DESCRIPTION
**Pull request by seb-hyland**
chaosDNA passes cargo run without errors, and should be ready for main. Please advise if there are any issues.

### CODE REVIEW

**Comment by seb-hyland** [comment]
I removed the chaosdna module from `main.rs` for now and left it as a comment, now checks are passing. I believe there's something weird going on with Git's file caching, and it's not recording the filenames properly. For now, chaosdna should be OK to merge into main.

**Comment by seb-hyland** [comment]
New commit: replaces insert_base with delete_base (fixed to match literature review)

File: src-tauri/src/chaodna.rs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		}
+            })
+	    .collect()
+    }
+
+    fn full_chaos(&self, initial_seq: &Vec<Base>, mod_prob: f64, del_prob: f64) -> Vec<Base> {
+	let mod_seq = self.modify_base(initial_seq, mod_prob);
+	self.delete_base(&mod_seq, del_prob)
+    }
+
```
Comment (lhao03) [inline src-tauri/src/chaodna.rs:62]: For full_chaos, I was thinking to simulate edit errors, we define $P_{edit}$ as the probability that an edit error occurs, then, we have $P_{edit} = P_{mutation} + P_{delete}$. The code may look something like this:

TLDR: While we map over a sequence of bases, we first randomly sample, does an edit error occur at all? And if it does, we randomly sample the type of edit error.
 
1. there is a probability that an edit error occurs, if edit error doesn't occur, we continue to the next base, if an edit error does occur, then we now have two discrete types of edit errors that can occur.
2. we randomly sample the two discrete types of errors, and then if deletion error occurs, we return no base, if mutation occurs, then we randomly sample again, which mutation occurs (should this be only of three bases of all four?).

The way your `full_chaos` is set up, we first randomly mutate, and then randomly delete; I'm not sure if this is the same definition of $P_{edit}$, and we might "delete" a mutated base, which I don't think would occur in synthesis? 

Let's discuss on Discord and in our software design meeting!

File: src-tauri/src/main.rs [hunk 2]
```diff
@@ -7,6 +7,7 @@ mod compressor;
 mod encoder;
 mod primer;
 mod sequence;
+// mod chaosdna;
```
Comment (lhao03) [inline src-tauri/src/main.rs:10]: the mod has to match the file name, which right now is `chaodna.rs`, not `chaosdna.rs`

**Review by lhao03 — CHANGES_REQUESTED** [review]
I left some thoughts on the implementation of edit errors, we can discuss on Discord/Meeting!

File: src-tauri/src/chaodna.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+// Here is some example code that I used to test:
+
+// fn main() {
+//     let chaos = Chaos{};
+//     let initial_sequence = vec![Base::G, Base::T, Base::A, Base::C, Base::C, Base::G, Base::A, Base::T, Base::T, Base::G];
+//     let modified_sequence = chaos.full_chaos(&initial_sequence, 0.2, 0.2);
+//     println!("{:?}", modified_sequence);
+// }
```
Comment (lhao03) [inline src-tauri/src/chaodna.rs:76]: You can add tests like this:

```rust
#[cfg(test)]
mod tests {
    use crate::chaosdna::{Chaos, RandError};
    use crate::primer::Base

    fn test_mutate() {
        let chaos = Chaos {};
        let initial_sequence = vec![
            Base::G,
            Base::T,
            Base::A,
            Base::C,
            Base::C,
            Base::G,
            Base::A,
            Base::T,
            Base::T,
            Base::G,
        ];
        let modified_sequence = chaos.full_chaos(&initial_sequence, 0.2, 0.2);
        println!("{:?}", modified_sequence);
    }
}
```

Some examples: 
https://doc.rust-lang.org/book/ch11-01-writing-tests.html

https://doc.rust-lang.org/rust-by-example/testing/unit_testing.html

**Review by lhao03 — CHANGES_REQUESTED** [review]
some documentation for tests

File: src-tauri/src/chaodna.rs (hunk 1 above)
Comment (seb-hyland) [reply src-tauri/src/chaodna.rs:62]: Sure! Just to be clear, assuming $P_{edit}$ occurs, what would the probability of deletion/mutation be set to? Would those be defined in input? Would the input then be vector, $P_{edit}$, $P_{del}$, $P_{mut}$?

As for the order of `full_chaos`, this was an oversight on my part; I initially had it call `modify_base` and then `insert_base` after; I later swapped all instances of `insert_base` with `delete_base` when I rewrote the function and forgot to change the order.

File: src-tauri/src/main.rs (hunk 2 above)
Comment (seb-hyland) [reply src-tauri/src/main.rs:10]: 🤦 I can't believe I didn't notice that lol

File: src-tauri/src/chaodna.rs (hunk 1 above)
Comment (lhao03) [reply src-tauri/src/chaodna.rs:62]: Yeah, those $P_{mut}$, $P_{del}$ would be defined, and after your modify your implementation then I think it will make sense.

File: src-tauri/src/main.rs (hunk 2 above)
Comment (seb-hyland) [reply src-tauri/src/main.rs]: Fixed in latest commits

File: src-tauri/src/chaodna.rs (hunk 1 above)
Comment (lhao03) [reply src-tauri/src/chaodna.rs]: Sorry, $P_{edit}$, $P_{del}$ and $P_{mut}$ would all be defined. Let me know if this is unclear

Comment (lhao03) [reply src-tauri/src/chaodna.rs]: image

image

Via [Short Systematic Codes for Correcting Random Edit Errors in DNA Storage](https://www.notion.so/ubcigem/Short-Systematic-Codes-for-Correcting-Random-Edit-Errors-in-DNA-Storage-cfd45e2b2a504c8f83494fad6747a91a?pvs=4)

File: src-tauri/src/chaodna.rs (hunk 3 above)
Comment (lhao03) [reply src-tauri/src/chaodna.rs]: Please write tests instead of running the file individually. Thank you!

File: src-tauri/src/chaodna.rs (hunk 1 above)
Comment (seb-hyland) [reply src-tauri/src/chaodna.rs]: Ah I see. So $P_{edit}$ defines the error probability, and then $P_{del/mut}$ define the skew between deletion and mutation. What happens if the latter two don’t sum to 1? Should the program continue or should it return an error? If it should continue, then there is a chance that $P_{edit}$ occurs but there is no change…

Alternatively, we could just have one parameter for $P_{del}$ and then implicitly calculate $P_{mut}$ as $1-P_{del}$. In GUI there could be a slider to adjust the relative proportion of these two.

Comment (lhao03) [reply src-tauri/src/chaodna.rs]: We can implicitly calculate $P_{mut}$ based on $P_{del}$ and show a slider in the GUI

Comment (seb-hyland) [reply src-tauri/src/chaodna.rs]: I have pushed up this implementation to the branch.

File: src-tauri/src/chaodna.rs (hunk 3 above)
Comment (seb-hyland) [reply src-tauri/src/chaodna.rs]: Done with latest commit

File: src-tauri/src/chaosdna.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	println!();
+	
+        let modified_sequence = chaos.full_chaos(&initial_sequence, 0.3, 0.5);
+	println!("Modified sequence: ");
+	for base in &modified_sequence {
+	    print!("{}", base);
+	}
+	println!();
+    }
+}
```
Comment (lhao03) [inline src-tauri/src/chaosdna.rs:76]: Last ask before merge; sorry I just thought of this. Can you make some sanity check tests?

1. $P_{edit} = 0$, so no matter what $P_{mut}$ is, the sequence that is returned is not mutated, in other words, the output sequence is the exact same as the input sequence.
2. $P_{mut} = 0$, the sequence that is returned is either the exact same or has bases deleted (because $P_{del} = 1$ (length must be shorter).
3. $P_{mut} = 1$, the sequence that is returned either has exact same length as the original sequence (because $P_{del} = 0$.

Thank you for working on this!

**Review by lhao03 — CHANGES_REQUESTED** [review]

File: src-tauri/src/chaosdna.rs (hunk 4 above)
Comment (seb-hyland) [reply src-tauri/src/chaosdna.rs]: Hi! With my latest commit I added 3 tests and additional in-code documentation.

I'm not sure exactly how to check 2; instead, I ran it with $P_{del}$ as 1 and $P_{edit}$ as 1 so that all bases are deleted, and checked that the length is 0. Let me know if there is a better way to test for what you are hoping for.

Please let me know if there are better naming conventions I can use; I'm not particularly enamored with my current function names (esp. for tests) so let me know if we want to use some standard naming convention for all tests.

Let me know if there's anything else I can add!

Comment (lhao03) [reply src-tauri/src/chaosdna.rs]: This works! Thanks! 😄

**Review by lhao03 — APPROVED** [review]
Thanks for working on this!

File: src-tauri/src/chaosdna.rs (hunk 4 above)
Comment (seb-hyland) [reply src-tauri/src/chaosdna.rs]: Great! Should I merge?

**Comment by lhao03** [comment]
Yep you can merge!

File: src-tauri/src/chaosdna.rs (hunk 4 above)
Comment (lhao03) [reply src-tauri/src/chaosdna.rs]: I replied on email but not sure if you got it; yep you can merge!
